target/linux/ath79/nand/base-files/lib/upgrade/glinet.sh

```diff
@@ -0,0 +1,33 @@
+#!/bin/sh
+
+# SPDX-License-Identifier: GPL-2.0
+#
+# Copyright (C) 2019 Jeff Kletsky
+#
+
+glinet_using_boot_dev_switch() {
+	if [ "$(fw_printenv -n boot_dev 2>/dev/null)" = "on" ] ; then
+		>&2 echo "NOTE: boot_dev=on; use switch to control boot partition"
+		true
+	else
+		false
+	fi
+}
+
+glinet_set_next_boot_nand() {
+	! glinet_using_boot_dev_switch &&	\
+		fw_setenv bootcount 0 &&	\
+		>&2 echo "Next boot set for NAND"
+}
+
+glinet_set_next_boot_nor() {
+	! glinet_using_boot_dev_switch &&	\
+		fw_setenv bootcount 3 &&	\
+		>&2 echo "Next boot set for NOR"
+}
+
+glinet_nand_nor_do_upgrade() {
+	set_next_boot_nand() { glinet_set_next_boot_nand; }
+	set_next_boot_nor() { glinet_set_next_boot_nor; }
+	nand_nor_do_upgrade "$1"
```

Comment:
Hmm, you are using a function here that is in the file sourcing this file. So, glinet.sh can not be included elsewhere, unless there is a function nand_nor_do_upgrade defined in that file. (To be honest, this is mainly a style issue and not a practical problem, though.)
Still, I don't really think that's desirable. However, I will leave the verdict about that to the person properly reviewing this section.

Reply:
There is no reason to include `glinet.sh` anywhere else. It provides the GL.iNet-specific  "flavor" of NAND/NOR selection, which is highly dependent on the specifics of the U-Boot installed on the device. GL.iNet provides a highly customized, full-featured U-Boot.

`nand_nor_do_upgrade()` is defined in `target/linux/ath79/nand/base-files/lib/upgrade/platform.sh` and references the generic `set_next_boot_nand()` and `set_next_boot_nor()` 

This provides a generic framework for devices that follow to be able to leverage the common `nand_nor_do_upgrade()` without repeating its general flow, simply by defining proper functions for `set_next_boot_nand()` and `set_next_boot_nor()` 

By placing the GL.iNet definitions in their own, manufacturer-specific file, a future enhancement of only installing it on devices that can use it is straightforward, such as done with `package/firmware/ipq-wifi`. There is precedent for the separation in `target/linux/ath79/generic/base-files/lib/functions/k2t.sh`, which could be similarly only installed on what appears to be only a single device that utilizes that code.